SES-4433 : Handle landscape mode

jbsession opened the pull request:
This PR removes portrait lock.

This PR also adds some new composables for different orientations.

TODO: Full end to end testing for landscape.

ThomasSession commented on app/src/main/java/org/thoughtcrime/securesms/giph/ui/compose/GiphyFragmentCompose.kt:
```diff
@@ -0,0 +1,47 @@
+        modifier = Modifier.fillMaxWidth().padding(24.dp),
+        contentAlignment = Alignment.Center
+    ) {
+        Text(text = stringResource(messageId), style = MaterialTheme.typography.bodyMedium)
```
We don't reference material's typography in the app, but instead use our own type system.
Instead of `MaterialTheme.typography.bodyMedium` this should use `LocalType.current.large` (because `large` matches the 16sp that were used in the xml file).

jbsession replied:
Updated.

ThomasSession commented on app/src/main/java/org/thoughtcrime/securesms/giph/ui/GiphyActivity.java:
```diff
@@ -77,15 +79,25 @@ private void initializeResources() {
-    }).attach();
+//    new TabLayoutMediator(binding.tabLayout, binding.giphyPager, (tab, position) -> {
+//      tab.setText(position == 0 ? NonTranslatableStringConstants.GIF : getString(R.string.stickers));
+//    }).attach();
```
This can be deleted if we are no longer using it

jbsession replied:
Updated.

ThomasSession commented on app/src/main/java/org/thoughtcrime/securesms/giph/ui/compose/GiphyTabsCompose.kt:
```diff
@@ -0,0 +1,116 @@
+    selectedIndex: Int,
+    onSelect: (Int) -> Unit
+) {
+    val indicatorColor = colorFromAttrOr(
```
We shouldn't rely on material themes directly in the app.
We already have a `SessionTabRow` that is styled as per the rest of the app, so it should be used here.

jbsession replied:
Updated.

ThomasSession commented on app/src/main/java/org/thoughtcrime/securesms/giph/ui/compose/GiphyTabsCompose.kt:
```diff
@@ -0,0 +1,116 @@
+}
+
+@Composable
+private fun colorFromAttrOr(@AttrRes attrResId: Int, fallback: Color): Color {
```
This likely can be deleted as we already have a styled tab component, `SessionTabRow`

jbsession replied:
Updated.

ThomasSession commented on app/src/main/java/org/thoughtcrime/securesms/home/startconversation/home/StartConversation.kt:
```diff
@@ -65,100 +78,186 @@ internal fun StartConversationScreen(
+        modifier = Modifier.verticalScroll(rememberScrollState())
+    ) {
+        ActionList(navigateTo = navigateTo)
+        Column(
```
Isn't there a way for both landscape and portrait to reuse some element instead of redeclaring them here?

jbsession replied:
Updated, I made the QrPanel reusable.

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/giph/ui/compose/GiphyFragmentCompose.kt:
```diff
@@ -0,0 +1,47 @@
+import network.loki.messenger.R
+
+/** Called from Java to set Compose content for the loading view. */
+fun setGiphyLoading(composeView: ComposeView) {
```
I updated this one to use a state now instead of redeclaring and resetting it every time. I think that made it more efficient?

jbsession commented:
> Here are a few issues from testing the branch:
> 
> * The horizontal paddings are not respected when in landscape mode. The app extends below the three dot menu on the left and below the status bar on the right. We should be handling that properly as you can otherwise lose some content.
>   In the homescreen for example, the dates are cut off on the right.
>   I'm not sure if we should allow content directly below the inset either. Google apps seems to leave a gap on the right hand side to cater for the camera inset space. It's quite jarring in a conversation for example, where the message bubble and content is hidden below the inset.
>   You also can't hit the back button from the top app bar in a conversation as it is hidden below the "three button" menu from the system.
> * `android:screenOrientation="portrait"` is still present on the webrtc call activity. With android 16 this type of call won't be allowed so we need to handle landscape in there accordingly and make sure it works fine, including with the camera on and the phone changing from portrait to landscape as there are some hardcoded logic in the screen for this, which auto rotates the icons based on orientation.
> * Viewing the QR code in recovery password creates a very large qr code that can't be viewed properly. It should probably try to scale itself like in other places, like the qr code acceissible from the qr icon in the top right of the settings.
> * The "create group" bottomsheet has very little space for the user list, though I'm not sure we can do much about this.  We could have the title and search also scroll with the rest to give more space, but that's something we'd have to discuss with the team, as it might not be as good of a behaviour for portrait where it's nice to only scroll the users and not the rest of the content.
> * The seesion network page looks a bit ugly with the giant node icon. Not a huge deal breaker but we could optimise it.

Updates on this:

- Insets are fixed in home, conversation screens. Also how do we feel about the camera insets?  (Other screens currently in progress)
- Removed the protrait lock for WebRtc call activity
- Updated the large QR code in recovery password

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/ui/adaptive/AdaptiveLayout.kt:
```diff
@@ -0,0 +1,63 @@
+import androidx.compose.ui.platform.LocalConfiguration
+
+object AdaptiveBreakpoints {
+    const val TWO_PANE_LANDSCAPE_MIN_WIDTH_DP = 480
```
I removed the initial two pane content here and kept the landscape first. I'll add to this helper in the future when we decide to cater for multiple screen sizes.

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/preferences/QRCodeActivity.kt:
```diff
@@ -127,3 +161,39 @@ fun QrPage(string: String) {
+            .padding(horizontal = LocalDimensions.current.mediumSpacing)
+    ) {
+        // Scale QR to the shorter side to avoid overflow in landscape. Clamp for sanity
+        val shortest: Dp = if (maxWidth < maxHeight) maxWidth else maxHeight
```
Updated.

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/home/startconversation/home/StartConversation.kt:
```diff
@@ -65,100 +78,177 @@ internal fun StartConversationScreen(
+        modifier = modifier
+    ) {
+        val qrModifier = if (getAdaptiveInfo().isLandscape) {
+            val shortest: Dp = if (maxWidth < maxHeight) maxWidth else maxHeight
```
Updated.

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/home/startconversation/home/StartConversation.kt:
```diff
@@ -65,100 +78,177 @@ internal fun StartConversationScreen(
+    ) {
+        val qrModifier = if (getAdaptiveInfo().isLandscape) {
+            val shortest: Dp = if (maxWidth < maxHeight) maxWidth else maxHeight
+            val qrSide = (shortest * 0.70f).coerceIn(160.dp, 520.dp)
```
So the values in the `coerceIn` are the clamps for the image. There are no actual rules for these, however these values prevent the image from being comically small or large (for tablets mostly). 
There are also no exact rules for the fractions such `70f` for the QR image. Currently this value works great for all landscape screen sizes.
I believe these values are product decisions that can be discussed with the team. These are initial working values for now. For a bit more readability I added the values in the Dimensions class.

ThomasSession commented on app/src/main/java/org/thoughtcrime/securesms/webrtc/CallManager.kt:
```diff
@@ -751,7 +751,9 @@ class CallManager @Inject constructor(
 
-    fun setDeviceOrientation(orientation: Orientation) {
+    fun setDeviceOrientation(orientation: Orientation, isAutoRotateOn : Boolean) {
+        if(isAutoRotateOn) return // let the activity handle the rotation
```
Why pass the `isAutoRotateOn` value here at all? This class shouldn't need to be aware of this.
The ViewModel could decide not to call this directly based on the value it is given by the activity

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/home/HomeActivity.kt:
```diff
@@ -585,75 +589,98 @@ class HomeActivity : ScreenLockActionBarActivity(),
+        val address = Address.fromSerialized(addressStr)
+
+        // Fetch the current ThreadRecord for this address
+        val thread = threadDb.getThreads(listOf(address)).firstOrNull() ?: return
```
Will update this to get the ThreadRecord from `data` in the viewmodel then.

ThomasSession replied:
I think this is a situation where you need to look at who is using this and ask yourself: Do we really need to pass a threadRecord everywhere?
Most of the places using this are probably only using an address, which you already have from the sheet's arguments

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/home/ConversationOptionsBottomSheet.kt:
```diff
@@ -218,5 +260,26 @@ class ConversationOptionsBottomSheet(private val parentContext: Context) : Botto
+        val sheet = dlg.findViewById<android.widget.FrameLayout>(R.id.design_bottom_sheet)
+            ?: return
+
+        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE){
```
This is probably because items in the home search has 2 items in it and having it set to Half expanded shows only 1 item. Setting Peek height might be better.

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/webrtc/CallViewModel.kt:
```diff
@@ -89,10 +89,7 @@ class CallViewModel @Inject constructor(
         get() = callManager.videoState
 
     var deviceOrientation: Orientation = Orientation.UNKNOWN
```
It is still used in the Rt call activity.

ThomasSession commented on app/src/main/java/org/thoughtcrime/securesms/home/startconversation/home/StartConversation.kt:
```diff
@@ -65,100 +82,181 @@ internal fun StartConversationScreen(
+    modifier: Modifier = Modifier
+) {
+    Column(
+        modifier = modifier.widthIn(max = 420.dp),
```
You can use your `smallMaximumImgClamp` which maybe can be renamed to something more meaningful since it doesn't apply only to images, but to content. `maxContentSizeSmall` ? Which means the other ones would need to be renamed as well

ThomasSession commented:
Some issues from the latest changes:
- App crashes when going from landscape to portrait in the camera screen
- Media rail is now disappeared when opening a set of images  inside a conversation, in landscape

ThomasSession commented on app/src/main/java/org/thoughtcrime/securesms/util/ViewUtilities.kt:
```diff
@@ -154,7 +154,11 @@ fun View.applySafeInsetsMargins(
     typeMask: Int = WindowInsetsCompat.Type.systemBars() or WindowInsetsCompat.Type.ime(),
-    additionalInsets : Insets = Insets.NONE // for additional offsets
+    additionalInsets : Insets = Insets.NONE, // for additional offsets
+    applyTop: Boolean = true,
```
I feel like this is a bit applying a band aid on top of a band aid.
We already have a method for setting up insets but now we're adding flags to cancel the act we are setting ourselves.
I think the logic is breaking at a higher level.
Is the issue that we are applying a bottom padding to the FAB which then gets sets on the side? If so which should fix that logic rather than adding more patches to the existing function.

jbsession replied:
Makes sense. The issue is that insets are applied to the sides as well which was intended for this method. For this case, maybe adding a new helper is better?

jbsession replied:
Yup, I can revert this method. Maybe forcing this method for this Fab is not the right approach since it only ever wants the bottom inset. That was the reason I added such flags in the first place.  I can then add a new helper to use for the case of this FAB, wdyt?

jbsession commented on app/src/main/java/org/thoughtcrime/securesms/mediasend/MediaSendActivity.kt:
```diff
@@ -498,6 +457,74 @@ class MediaSendActivity : ScreenLockActionBarActivity(), MediaPickerFolderFragme
+                    handleCameraImageCaptured(uri, size, width, height)
+                }
+            }else{
+                if(supportFragmentManager.backStackEntryCount == 0){
```
This will avoid a black screen in the MediaSendActivity finish it for when we press back to "undo" a photo, and then pressing cancelling the CameraActivity.
